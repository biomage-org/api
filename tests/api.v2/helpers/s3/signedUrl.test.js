const signedUrl = require('../../../../src/api.v2/helpers/s3/signedUrl');

const { getSignedUrl, getSampleFileUploadUrl } = signedUrl;

const AWS = require('../../../../src/utils/requireAWS');

jest.mock('../../../../src/utils/requireAWS', () => ({
  S3: jest.fn(),
}));

describe('getSignedUrl', () => {
  const signedUrlSpy = jest.fn();

  const testParams = {
    Bucket: 'test-bucket',
    Key: 'test-key',
  };

  beforeEach(() => {
    AWS.S3.mockReset();
    AWS.S3.mockImplementation(() => ({
      getSignedUrl: signedUrlSpy,
    }));
  });

  it('Should call S3 signed url correctly', () => {
    getSignedUrl('getObject', testParams);
    expect(signedUrlSpy).toHaveBeenCalledWith('getObject', testParams);
  });

  it('Should add the region config if the requested url is doing upload', () => {
    const expectedConfig = {
      region: 'eu-west-1',
    };

    getSignedUrl('putObject', testParams);

    // Expect config to have region
    expect(AWS.S3).toHaveBeenCalledWith(expect.objectContaining(expectedConfig));
  });

  it('Should add the region config if the requested url is not doing upload too', () => {
    const expectedConfig = {
      region: 'eu-west-1',
    };

    getSignedUrl('getObject', testParams);

    // Expect config to have region
    expect(AWS.S3).toHaveBeenCalledWith(expect.objectContaining(expectedConfig));
  });

  it('Should throw an error if bucket is not defined', () => {
    expect(() => {
      getSignedUrl('test-bucket', { Key: 'test-key' });
    }).toThrow();
  });

  it('Should throw an error if key is not defined', () => {
    expect(() => {
      getSignedUrl('test-bucket', { Bucet: 'test-bucket' });
    }).toThrow();
  });
});

describe('getSampleFileUploadUrl', () => {
  const mockSampleFileId = 'mockSampleFileId';

  const signedUrlResponse = 'signedUrl';

  const signedUrlSpy = jest.fn();

  beforeEach(() => {
    signedUrlSpy.mockReturnValueOnce(signedUrlResponse);

    AWS.S3.mockReset();
    AWS.S3.mockImplementation(() => ({
      getSignedUrl: signedUrlSpy,
    }));
  });

  it('works correctly without metadata', () => {
    const response = getSampleFileUploadUrl(mockSampleFileId, {});

    expect(response).toEqual(signedUrlResponse);
    expect(signedUrlSpy).toMatchSnapshot();
  });

  it('works correctly with metadata cellrangerVersion', () => {
    const response = getSampleFileUploadUrl(mockSampleFileId, { cellrangerVersion: 'v2' });

    expect(response).toEqual(signedUrlResponse);
    expect(signedUrlSpy).toMatchSnapshot();
  });
});
