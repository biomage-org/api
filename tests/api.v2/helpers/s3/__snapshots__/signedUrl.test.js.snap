// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSampleFileUploadUrl works correctly with metadata cellrangerVersion 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "putObject",
      Object {
        "Bucket": "biomage-originals-test",
        "Expires": 3600,
        "Key": "mockSampleFileId",
      },
    ],
    Array [
      "putObject",
      Object {
        "Bucket": "biomage-originals-test",
        "Expires": 3600,
        "Key": "mockSampleFileId",
        "Metadata": Object {
          "cellranger_version": "v2",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": "signedUrl",
    },
    Object {
      "type": "return",
      "value": "signedUrl",
    },
  ],
}
`;

exports[`getSampleFileUploadUrl works correctly without metadata 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "putObject",
      Object {
        "Bucket": "biomage-originals-test",
        "Expires": 3600,
        "Key": "mockSampleFileId",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": "signedUrl",
    },
  ],
}
`;
